File: ocaml/basics/02_functions.ml
(* TOPIC: Functions | ocaml 02_functions.ml *)
(* Docs: https://ocaml.org/docs/values-and-functions *)

(* TODO 1: Basic function definitions — all functions are curried by default *)
(*   let add x y = x + y                   (* curried: add : int -> int -> int *)
     let add5 = add 5                       (* partial application: int -> int *)
     let double = ( * ) 2                   (* operator as function *)  *)

(* TODO 2: Anonymous functions *)
(*   let square = fun x -> x * x
     let apply f x = f x
     let result = apply (fun x -> x + 1) 41  *)

(* TODO 3: Recursive functions — must use let rec *)
(*   let rec factorial n =
       if n <= 1 then 1 else n * factorial (n - 1)
     let rec fib n =
       if n <= 1 then n else fib (n-1) + fib (n-2)  *)

(* TODO 4: Higher-order functions — implement from scratch *)
(*   let rec my_map f = function
       | [] -> []
       | x :: xs -> f x :: my_map f xs
     let rec my_filter p = function
       | [] -> []
       | x :: xs -> if p x then x :: my_filter p xs else my_filter p xs
     let rec my_fold_left f acc = function
       | [] -> acc
       | x :: xs -> my_fold_left f (f acc x) xs  *)

(* TODO 5: Labeled arguments — ~label: *)
(*   let greet ~name ~greeting = greeting ^ ", " ^ name ^ "!"
     greet ~name:"Alice" ~greeting:"Hello"
     greet ~greeting:"Hi" ~name:"Bob"    (* can reorder labeled args *)  *)

(* TODO 6: Optional arguments — ?label *)
(*   let connect ?(port=80) host =
       Printf.sprintf "Connecting to %s:%d" host port
     connect "example.com"               (* uses default port 80 *)
     connect ~port:443 "secure.com"      *)

(* TODO 7: Function composition *)
(*   let ( >> ) f g x = g (f x)          (* pipe: x |> f |> g *)
     let ( << ) f g x = f (g x)          (* compose: f ∘ g *)
     let process = String.trim >> String.lowercase_ascii >> String.length  *)

(* TODO 8: Tail recursion — avoids stack overflow for large n *)
(*   let factorial n =
       let rec go acc n = if n <= 1 then acc else go (acc * n) (n - 1) in
       go 1 n
     (* OCaml optimizes tail-recursive calls; non-tail fib would stack overflow for large n *)  *)

let () = print_endline "TODO: implement function exercises"

File: kotlin/basics/02_functions.kt
package basics
// TOPIC: Functions | kotlinc 02_functions.kt -include-runtime -d out.jar && java -jar out.jar
// Docs: https://kotlinlang.org/docs/functions.html

fun main() {
    // TODO 1: Default parameters — fun greet(name: String, greeting: String = "Hello")
    //   - Define greet() above main() with a default greeting
    //   - Call with one arg (uses default), then call with both args
    //   - Show that default params remove need for overloads

    // TODO 2: Named arguments — call with named params in any order
    //   - Call greet(greeting = "Hey", name = "Alice") in any order
    //   - Show how named args improve readability for boolean flags

    // TODO 3: Extension functions — fun String.isPalindrome(): Boolean
    //   - Define isPalindrome() as a String extension above main()
    //   - Call "racecar".isPalindrome() and "hello".isPalindrome()
    //   - Define a second extension: fun List<Int>.secondOrNull(): Int?

    // TODO 4: Higher-order functions — fun transform(list: List<Int>, fn: (Int) -> Int): List<Int>
    //   - Define transform() above main()
    //   - Call it with a lambda: transform(listOf(1,2,3)) { it * 2 }
    //   - Pass a function reference: transform(list, ::square)

    // TODO 5: Lambda with receiver — fun buildString(block: StringBuilder.() -> Unit): String
    //   - Define buildString() above main() (note: stdlib has one, define your own)
    //   - Use it: val s = myBuildString { append("Hello"); append(" World") }
    //   - Explain how the receiver (this = StringBuilder) is available inside the block

    // TODO 6: Inline functions — inline fun measure(block: () -> Unit): Long
    //   - Define measure() that records System.currentTimeMillis() before/after block
    //   - Use it: val elapsed = measure { /* some work */ }
    //   - Explain why inline avoids lambda allocation overhead

    // TODO 7: Operator overloading — data class Vector(val x: Int, val y: Int), operator fun plus
    //   - Define Vector data class above main()
    //   - Add operator fun plus(other: Vector): Vector
    //   - Add operator fun times(scalar: Int): Vector
    //   - Use: val v3 = Vector(1,2) + Vector(3,4)

    // TODO 8: Tail-recursive functions — tailrec fun factorial(n: Long, acc: Long = 1): Long
    //   - Define tailrec factorial() above main()
    //   - Call factorial(20L) and print result
    //   - Explain: tailrec compiles to a loop, preventing stack overflow
    //   - Also implement tailrec fibonacci(n: Int, a: Int = 0, b: Int = 1): Int
}
